### ISSUE TITLE
Stateful Version

### ISSUE THREAD

**Issue by kayuksel** [tejaslodaya/timeseries-clustering-vae#5]
Hello, thank you for sharing this good work. Do you also plan to include a stateful version?

**Comment by kayuksel** [comment]
I assumed that it was Stateless as your training data as shuffled but then I also noticed that you are not resetting the state of LSTM. This seems like a contradiction to me. Can you please advise?

**Comment by hellojinwoo** [comment]
This issue is legitimate. I agree with kayuksel

**Comment by hellojinwoo** [comment]
I changed the forward part of the class Encoder as follows : 

before
 
```    
def forward(self, x):
  """Forward propagation of encoder. Given input, outputs the last hidden state of encoder

  :param x: input to the encoder, of shape (sequence_length, batch_size, number_of_features)
  :return: last hidden state of encoder, of shape (batch_size, hidden_size)
  """
  _, (h_end, c_end) = self.model(x)

  h_end = h_end[-1, :, :]
  return h_end
```

after

```    
def forward(self, x):
  """Forward propagation of encoder. Given input, outputs the last hidden state of encoder

  :param x: input to the encoder, of shape (sequence_length, batch_size, number_of_features)
  :return: last hidden state of encoder, of shape (batch_size, hidden_size)
  """

h_0 = Variable(torch.zeros(
            self.hidden_layer_depth, x.size(0), self.hidden_size))
        
 c_0 = Variable(torch.zeros(
            self.hidden_layer_depth, x.size(0), self.hidden_size))

   _, (h_end, c_end) = self.model(x, (h_0, c_0))

  h_end = h_end[-1, :, :]
  return h_end
```

Is this right?

**Comment by kayuksel** [comment]
Yes, that looks quite right to me. Except that in case of GRU, c0 should not be utilized:

`self.model(x, h0 if gru else (h0, c0))`

You also need to add requires_grad = False to Variable definitions and send them to device.

Do you also know why the author might have utilized h_end[0, :, :] in his version of the code?

**Comment by hellojinwoo** [comment]
@ kayuksel you should take a look at the issue right below yours. This one is talking about it, and it correts it from `h_end [0,:,:]` to `h_end[-1, : , :]`

**Comment by kayuksel** [comment]
@hellojinwoo Can you maybe commit these changes on your fork and then make a pull request?

**Comment by hellojinwoo** [comment]
@ kayuksel

I have one more question for you. Why do we need set the `requires_grad = False` to `h_0` and `c_0` ?

**Comment by kayuksel** [comment]
@hellojinwoo Never questioned the reason, that's the common practice I was using for a while now. However, I guess that since that the states are resetted after each forward; there is no need to calculate gradient for those and update them during the optimization step.

**Comment by hellojinwoo** [comment]
@kayuksel That makes sense, but I don't see `requires_grad = False` in other codes. But if you put the `requires_grad = False` there, wouldn't it be the problem that the `hidden_state` and `cell_state` is not updated at all during the training phase? It's because `requries_grad` makes these two states have no gradient_path.... It is my thought. Would appreciate if you clarify, thanks :D 

#### example 1 https://www.jessicayung.com/lstms-for-time-series-in-pytorch/

```
class LSTM(nn.Module):

    def __init__(self, input_dim, hidden_dim, batch_size, output_dim=1,
                    num_layers=2):
        super(LSTM, self).__init__()
        self.input_dim = input_dim
        self.hidden_dim = hidden_dim
        self.batch_size = batch_size
        self.num_layers = num_layers

        # Define the LSTM layer
        self.lstm = nn.LSTM(self.input_dim, self.hidden_dim, self.num_layers)

        # Define the output layer
        self.linear = nn.Linear(self.hidden_dim, output_dim)

    def init_hidden(self):
        # This is what we'll initialise our hidden state as
        return (torch.zeros(self.num_layers, self.batch_size, self.hidden_dim),
                torch.zeros(self.num_layers, self.batch_size, self.hidden_dim))
```

#### example 2 https://github.com/spdin/time-series-prediction-lstm-pytorch/blob/master/Time_Series_Prediction_with_LSTM_Using_PyTorch.ipynb

````

class LSTM(nn.Module):

    def __init__(self, num_classes, input_size, hidden_size, num_layers):
        super(LSTM, self).__init__()
        
        self.num_classes = num_classes
        self.num_layers = num_layers
        self.input_size = input_size
        self.hidden_size = hidden_size
        self.seq_length = seq_length
        
        self.lstm = nn.LSTM(input_size=input_size, hidden_size=hidden_size,
                            num_layers=num_layers, batch_first=True)
        
        self.fc = nn.Linear(hidden_size, num_classes)

    def forward(self, x):
        h_0 = Variable(torch.zeros(
            self.num_layers, x.size(0), self.hidden_size))
        
        c_0 = Variable(torch.zeros(
            self.num_layers, x.size(0), self.hidden_size))
        
        # Propagate input through LSTM
        ula, (h_out, _) = self.lstm(x, (h_0, c_0))
        
        h_out = h_out.view(-1, self.hidden_size)
        
        out = self.fc(h_out)
        
        return out
```

**Comment by kayuksel** [comment]
They would get updated only after an optimizer step and we set them back to
zero immediately in next forward; so in my opinion, that would not change
anything except for time required for calculation of their gradients.

**Comment by hellojinwoo** [comment]
I see your point! Thank you :D

**Comment by tejaslodaya** [comment]
Hi @kayuksel ,

Sorry for the late reply. I've been held up at work.

You're correct. This is already a stateful version.
I had developed this library at my work where we had client data and applied clustering on top of this. In order to mask that (NDA obviously), I had to pick a random open-source dataset.

No intentions of making it stateless (NN's are stateless, LSTM/GRU are stateful which is evident from my code).

**Comment by tejaslodaya** [comment]
Hi @hellojinwoo ,

To answer your 4th comment, you don't have to explicitly give h0 and c0.
Pytorch does that by default and initializes it and adds it to AutoGrad.

You needn't worry. The code i've written is already stateful

**Comment by tejaslodaya** [comment]
Let me know if it makes sense @kayuksel .
I'll close this issue.

**Comment by tejaslodaya** [comment]
@kayuksel add `required_grad=False` when you want a variable to act like a constant. In this case, if you're trying to build a stateful version, you have to include `required_grad=True` in order to let the network "learn" through backprop.

**Comment by kayuksel** [comment]
@tejaslodaya Yes, I am at the moment using a stateless version. Actually, the data that I have is suitable for the stateful but I don't have a previous experience with that. Not sure yet how I should feed data (daily balances of several accounts) in the stateful case. I should maybe check some more examples about that.

**Comment by tejaslodaya** [comment]
Thanks for the issue. Included it in FAQs section to increase visibility
